io/src/main/scala/fs2/io/net/Socket.scala

```diff
@@ -69,3 +76,140 @@ trait Socket[F[_]] {
     */
   def writes: Pipe[F, Byte, INothing]
 }
+
+object Socket {
+  private[net] def forAsync[F[_]: Async](
+      ch: AsynchronousSocketChannel
+  ): Resource[F, Socket[F]] =
+    Resource.make {
+      (Semaphore[F](1), Semaphore[F](1)).mapN { (readSemaphore, writeSemaphore) =>
+        new AsyncSocket[F](ch, readSemaphore, writeSemaphore)
+      }
+    }(_ => Async[F].delay(if (ch.isOpen) ch.close else ()))
+
+  private final class AsyncSocket[F[_]](
+      ch: AsynchronousSocketChannel,
+      readSemaphore: Semaphore[F],
+      writeSemaphore: Semaphore[F]
+  )(implicit F: Async[F])
+      extends Socket[F] {
+    private[this] final val defaultReadSize = 8192
+    private[this] var readBuffer: ByteBuffer = ByteBuffer.allocateDirect(defaultReadSize)
+
+    private def getBufferForRead(size: Int): F[ByteBuffer] = F.delay {
+      if (readBuffer.capacity() < size)
+        readBuffer = ByteBuffer.allocateDirect(size)
+      else
+        (readBuffer: Buffer).limit(size)
+      readBuffer
+    }
+
+    /** Performs a single channel read operation in to the supplied buffer. */
+    private def readChunk(buffer: ByteBuffer): F[Int] =
+      F.async_[Int] { cb =>
+        ch.read(
+          buffer,
+          null,
+          new IntCallbackHandler(cb)
+        )
+      }
+
+    /** Copies the contents of the supplied buffer to a `Chunk[Byte]` and clears the buffer contents. */
+    private def releaseBuffer(buffer: ByteBuffer): F[Chunk[Byte]] =
+      F.delay {
+        val read = buffer.position()
+        val result =
+          if (read == 0) Chunk.empty
+          else {
+            val dest = ByteBuffer.allocateDirect(read)
+            (buffer: Buffer).flip()
+            dest.put(buffer)
+            (dest: Buffer).flip()
+            Chunk.byteBuffer(dest)
+          }
+        (buffer: Buffer).clear()
+        result
+      }
+
+    def read(max: Int): F[Option[Chunk[Byte]]] =
+      readSemaphore.permit.use { _ =>
+        getBufferForRead(max).flatMap { buffer =>
+          readChunk(buffer).flatMap { read =>
+            if (read < 0) F.pure(None)
+            else releaseBuffer(buffer).map(Some(_))
+          }
+        }
+      }
+
+    def readN(max: Int): F[Option[Chunk[Byte]]] =
+      readSemaphore.permit.use { _ =>
+        getBufferForRead(max).flatMap { buffer =>
+          def go: F[Option[Chunk[Byte]]] =
+            readChunk(buffer).flatMap { readBytes =>
+              if (readBytes < 0 || buffer.position() >= max)
+                releaseBuffer(buffer).map(Some(_))
+              else go
+            }
+          go
```

Comment:
Since this recursion can only end up returning a `Some`, could we remove the `Option` wrapper and change the type to `F[Chunk[Byte]]`?